### PR TITLE
copr: support time

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @coocood @shenli @qiuyesuifeng PTAL

### CODE REVIEW

**Comment by shenli** [comment]
@tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM 
@coocood

**Comment by coocood** [comment]
LGTM

File: src/lib.rs [hunk 1]
```diff
@@ -50,6 +51,7 @@ extern crate hyper;
 #[cfg(unix)]
 extern crate nix;
 extern crate alloc;
+extern crate chrono;
```
Comment (siddontang) [inline src/lib.rs]: can we use this crate to replace time crate later.
time crate is deprecated now.

Comment (BusyJay) [inline src/lib.rs]: I'm not sure. Though `Time` is deprecated, but it just works fine. And this crate depends on time too.

File: src/util/codec/datum.rs [hunk 2]
```diff
@@ -301,8 +349,10 @@ impl Datum {
         }
     }
 
+    /// Keep compatible with TiDB's `ToDecimal` function.
     pub fn into_dec(self) -> Result<Decimal> {
         match self {
+            Datum::Time(t) => t.to_decimal().map_err(From::from),
```
Comment (siddontang) [inline src/util/codec/datum.rs]: can we use `try!` here directly instead of map_err ?

Comment (BusyJay) [inline src/util/codec/datum.rs]: Of cause, but `try` require two line, `map_err` only use one line.

File: src/util/codec/mysql/mod.rs [hunk 3]
```diff
@@ -11,10 +11,70 @@
...
+    if fsp > MAX_FSP {
+        return Err(invalid_type!("Invalid fsp {}", fsp));
+    }
+    Ok(())
+}
+
+/// Parse string as if it's a fraction part of a number and keep
+/// only `fsp` precision.
+fn parse_frac(s: &[u8], fsp: u8) -> Result<u32> {
+    if s.is_empty() {
```
Comment (siddontang) [inline src/util/codec/mysql/mod.rs]: no need to check fsp <= 6?

File: src/util/codec/mysql/time.rs [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+use chrono::{NaiveDate, NaiveDateTime, Timelike, Datelike, Duration};
+
+use util::codec::mysql::{self, types, parse_frac};
+use util::codec::mysql::Decimal;
+use util::codec::{Error, Result, TEN_POW};
+
+
+const ZERO_DATETIME_STR: &'static str = "0000-00-00 00:00:00";
+const ZERO_DATE_STR: &'static str = "0000-00-00";
+const ZERO_TIMESTAMP: i64 = -62169984000;
```
Comment (siddontang) [inline src/util/codec/mysql/time.rs]: why -62169984000 ? any comment?

File: src/lib.rs (hunk 1 above)
Comment (siddontang) [inline src/lib.rs]: got it.

File: src/util/codec/mysql/mod.rs (hunk 3 above)
Comment (BusyJay) [inline src/util/codec/mysql/mod.rs]: No, if `s` is empty, no matter what `fsp` is, result is always `0`.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**